### ISSUE TITLE
AddUserResponse fails to parse if no Country Code

### ISSUE THREAD

**Issue by jackBonadies** [jpdillingham/Soulseek.NET#613]
Hi JP,

This one is fortunately straightforward (I think).. The AddUserReponse message fails to parse correctly in the case of no Country Code and throws (tries to read data when there isn't any).  I don't know on which clients a Country Code is not implemented (other than my own client which is how I found this bug) but in the Nicotine protocol docs (https://nicotine-plus.github.io/nicotine-plus/doc/SLSKPROTOCOL.html#server-code-5) regarding country code it does state "(may not be implemented)" and in the Nicotine source code it does a check to see if there is remaining data before reading.  In the potential fix I made, I set the country code to an empty string if there is none.  This is a safe default since it is the string sent for users without a country code (i.e. on QT, go into a chatroom, those users with an green-ish Owl rather than a flag)

I made the potential fix here: https://github.com/jackBonadies/SeekerAndroid/commit/656597edea228f37df824d02aa2b3528fc51ed4c
Let me know and if you want I can make a pull request.

If you want to test this case you can feel free to use a client device I have with username "7fortress" which does not send a country code.

Thanks!

**Comment by jpdillingham** [comment]
Thanks for pointing this out!  I'm guessing it isn't a problem for `UserJoinedRoomNotification`, or I would have seen it by now, but could you join a channel with this user and confirm?

I'd gladly accept a PR if you'd like to make one.  Your fix looks good to me, other than I'd prefer that `CountryCode` defaults to null.

Here's a test that will maintain coverage:

```c#
        [Trait("Category", "Parse")]
        [Theory(DisplayName = "Parse does not throw if CountryCode is missing"), AutoData]
        public void Parse_Does_Not_Throw_If_CountryCode_Is_Missing(string username, UserData userData)
        {
            var msg = new MessageBuilder()
                .WriteCode(MessageCode.Server.AddUser)
                .WriteString(username)
                .WriteByte(1) // exists = true
                .WriteInteger((int)userData.Status)
                .WriteInteger(userData.AverageSpeed)
                .WriteLong(userData.UploadCount)
                .WriteInteger(userData.FileCount)
                .WriteInteger(userData.DirectoryCount)
                .Build();

            var r = AddUserResponse.FromByteArray(msg);

            Assert.Equal(username, r.Username);
            Assert.True(r.Exists);
            Assert.Equal(userData.Status, r.UserData.Status);
            Assert.Equal(userData.AverageSpeed, r.UserData.AverageSpeed);
            Assert.Equal(userData.UploadCount, r.UserData.UploadCount);
            Assert.Equal(userData.FileCount, r.UserData.FileCount);
            Assert.Equal(userData.DirectoryCount, r.UserData.DirectoryCount);
            Assert.Null(r.UserData.CountryCode);
        }
```

**Comment by jackBonadies** [comment]
I will make a pull request with Country Code set to null rather than empty string, and with said test added.  

After doing some more testing I found the following regarding the server's response to AddUser (some of which was contrary to my initial comments):
If you add a user that exists but is offline, the response to AddUser will not have a country code (causing the no data bug).
If one is on wifi (has an IPv4 address), the country code will be valid (i.e. "US").
If one is on data (has an IPv6 address), the country code will be an empty string (i.e. "") in this case it parses just fine.

I think the above might be client independent.. I am not sure though, the last two observations are just on my own client.

This is not a problem with UserJoinedRoomNotification, in my testing, joining a room with IPv6 got a UserJoinedRoomNotification with an empty string for country code.  Since one cannot both be offline and join a room, I don't think the no data parse bug can occur.  I did look at Nicotine source and saw that it did check for no data on UserJoinedRoomNotification, but whether this condition is actually possible, I am not sure - based on my above experiments I don't think it is.

**Comment by mathiascode** [comment]
> I did look at Nicotine source and saw that it did check for no data on UserJoinedRoomNotification, but whether this condition is actually possible, I am not sure - based on my above experiments I don't think it is.

I believe most of the checks in Nicotine+ were added many years ago to maintain compatibility with [Soulfind](https://github.com/seeschloss/soulfind), which has been unmaintained since 2006.

**Comment by jpdillingham** [comment]
Closed by #614 

Thanks again @jackBonadies !
